storage/rocksdb/clone/common.h

```diff
@@ -0,0 +1,697 @@
+/*
+  Copyright (C) 2022, Laurynas Biveinis
+
+  This program is free software; you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation; version 2 of the License.
+
+  This program is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with this program; if not, write to the Free Software
+  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA */
+
+#ifndef CLONE_COMMON_H_
+#define CLONE_COMMON_H_
+
+#include <sys/types.h>
+#include <atomic>
+#include <cstddef>
+#include <cstdint>
+#include <ctime>
+#include <functional>
+#include <string>
+#include <unordered_set>
+
+#include "my_dbug.h"
+#include "my_inttypes.h"
+#include "mysql/plugin.h"
+#include "mysql/psi/mysql_cond.h"
+#include "mysql/psi/mysql_mutex.h"
+#include "mysql/psi/mysql_rwlock.h"
+#include "mysqld_error.h"
+
+#include "../rdb_buff.h"
+#include "../rdb_psi.h"
+
+namespace myrocks {
+
+namespace clone {
+
+constexpr char checkpoint_name_prefix[] = ".clone_checkpoint-";
+constexpr char in_place_temp_datadir[] = ".rocksdb.cloned";
+constexpr char in_place_temp_wal_dir[] = ".rocksdb.wal.cloned";
+constexpr char in_progress_marker_file[] = ".clone_in_progress";
+
+constexpr auto dir_mode = S_IRWXU;
+
+constexpr std::uint64_t invalid_file_id =
+    std::numeric_limits<std::uint64_t>::max();
+constexpr my_off_t invalid_file_offset = std::numeric_limits<my_off_t>::max();
+constexpr char invalid_file_name[] = "";
+
+// TODO(laurynas): use outside clone too
+inline std::string file_in(const std::string &dir, const std::string &fn) {
+  std::string result;
+  result.reserve(dir.length() + fn.length() + 2);  // FN_DIRSEP and \0
+  result = dir;
+  result += FN_DIRSEP;
+  result += fn;
+  return result;
+}
+
+enum class metadata_type : std::uint32_t {
+  LOCATOR_V1,
+  FILE_NAME_MAP_V1,
+  FILE_CHUNK_V1,
+  ADD_ESTIMATE_V1,
+  MAX_VALID_TYPE = ADD_ESTIMATE_V1,
+  INVALID
+};
+
+class [[nodiscard]] metadata_header final {
```

Comment:
Could you add more comments about the metadata header usage, when version is expected to be bumped up?

Reply:
Added a header comment here, let me know if something's missing